### DISCUSSION TITLE
Send data to device (H2D)

### DISCUSSION THREAD

**Discussion by sonik-br** [hathach/tinyusb#2053] [Q&A] [answered]
I'm trying to send a byte to a device (from host to device, using pico as host).
I know endpoint address (0x01OUT/H2D).

It already exists this code. I think it uses libusb:
`unsigned char value = 0x0F;`
`unsigned int timeout = 10;`
`usb_control_msg(device, USB_ENDPOINT_OUT | USB_TYPE_VENDOR | USB_RECIP_INTERFACE, 1, value, 0, 0, 0, timeout);`

I'm trying to port this to tinyusb but I can't find any code sample for this.
Looks like `tuh_control_xfer` can do this but I don't know how to use.

Any advice?

**Comment by manuelbl** [comment]
You have to implement the function `tud_vendor_control_xfer_cb`.

The only example I can find is:

https://github.com/hathach/tinyusb/blob/5e023fa2ca4c20f06b6e0dc12f6f044a7d4e14bd/examples/device/webusb_serial/src/main.c#L169

**Comment by sonik-br** [comment]
I think the `tud_*` mehods are for device and not host. If I understand correctly how the lib names it's stuff. Host methods are `tuh_*`
I'm using the pico as a host. Looks like I need to claim and open the endpoint but I don't know how.
I will update the origial post to make it clear that I'm using the pico as a host :)

**Reply by manuelbl** [reply, marked as answer]
You are correct of course. I didn't read the question carefully and misunderstood the scenario.

I'm not really familiar with TinyUSB's host implementation but here some hints:

- For control requests, you don't need to claim or open anything. The control endpoint is always available, even if the control request is directed at an interface or endpoint.
- The *libusb* code likely translates to:

```C
unsigned char value = 0x0F;
tusb_control_request_t const request = {
    .bmRequestType_bit = {
        .recipient = TUSB_REQ_RCPT_INTERFACE,
        .type      = TUSB_REQ_TYPE_VENDOR,
        .direction = TUSB_DIR_OUT
    },
    .bRequest = 1,
    .wValue   = value,
    .wIndex   = 0,
    .wLength  = 0
};
tuh_xfer_t xfer = {
    .daddr       = daddr, // from mount callback
    .ep_addr     = 0, // control endpoint has address 0
    .setup       = &request,
    .buffer      = NULL,
    .complete_cb = complete_cb // your callback to be called when the request completes
};
tuh_control_xfer(&xfer);
```

- You will need to implement the timeout yourself.

**Comment by sonik-br** [comment]
Thank you very much!
Got it working.

Now I need to see how to also read data from a non-hid device :)
